tests/test_spi_devices.py

```diff
@@ -0,0 +1,339 @@
+
+
+class MockMCP3208(MockMCP3xxx):
+    def __init__(self, clock_pin, mosi_pin, miso_pin, select_pin=None):
+        super(MockMCP3208, self).__init__(
+            clock_pin, mosi_pin, miso_pin, select_pin, channels=8, bits=12)
+
+
+class MockMCP3301(MockMCP3xxx):
+    def __init__(self, clock_pin, mosi_pin, miso_pin, select_pin=None):
+        super(MockMCP3301, self).__init__(
+            clock_pin, mosi_pin, miso_pin, select_pin, channels=2, bits=12)
+
+    def on_start(self):
+        super(MockMCP3301, self).on_start()
+        result = self.channels[0] - self.channels[1]
+        result = clamp(result, -self.vref, self.vref)
+        result = scale(result, self.vref, self.bits)
+        if result < 0:
+            result += 8192
+        self.tx_word(result, self.bits + 4)
+
+
+class MockMCP3302(MockMCP33xx):
+    def __init__(self, clock_pin, mosi_pin, miso_pin, select_pin=None):
+        super(MockMCP3302, self).__init__(
+            clock_pin, mosi_pin, miso_pin, select_pin, channels=4)
+
+
+class MockMCP3304(MockMCP33xx):
+    def __init__(self, clock_pin, mosi_pin, miso_pin, select_pin=None):
+        super(MockMCP3304, self).__init__(
+            clock_pin, mosi_pin, miso_pin, select_pin, channels=8)
+
+
+def single_mcp_test(mock, pot, channel, bits):
+    scale = 2**bits
+    tolerance = 1 / scale
+    mock.channels[channel] = 0.0
+    assert pot.raw_value == 0
+    assert isclose(pot.value, 0.0, abs_tol=tolerance)
+    mock.channels[channel] = mock.vref / 2
+    assert pot.raw_value == (scale / 2) - 1
+    assert isclose(pot.value, 0.5, abs_tol=tolerance)
+    mock.channels[channel] = mock.vref
+    assert pot.raw_value == scale - 1
+    assert isclose(pot.value, 1.0, abs_tol=tolerance)
+
+def differential_mcp_test(mock, pot, pos_channel, neg_channel, bits, full=False):
+    scale = 2**bits
+    tolerance = 1 / scale
+    mock.channels[pos_channel] = 0.0
+    mock.channels[neg_channel] = 0.0
+    assert pot.raw_value == 0
+    assert isclose(pot.value, 0.0, abs_tol=tolerance)
+    mock.channels[pos_channel] = mock.vref / 2
+    assert pot.raw_value == (scale / 2) - 1
+    assert isclose(pot.value, 0.5, abs_tol=tolerance)
+    mock.channels[pos_channel] = mock.vref
+    assert pot.raw_value == scale - 1
+    assert isclose(pot.value, 1.0, abs_tol=tolerance)
+    mock.channels[neg_channel] = mock.vref / 2
+    assert pot.raw_value == (scale / 2) - 1
+    assert isclose(pot.value, 0.5, abs_tol=tolerance)
+    mock.channels[pos_channel] = mock.vref / 2
+    assert pot.raw_value == 0
+    assert isclose(pot.value, 0.0, abs_tol=tolerance)
+    mock.channels[pos_channel] = 0.0
+    mock.channels[neg_channel] = mock.vref
+    if full:
+        assert pot.raw_value == -scale
+        assert isclose(pot.value, -1.0, abs_tol=tolerance)
+    else:
+        assert pot.raw_value == 0
+        assert isclose(pot.value, 0.0, abs_tol=tolerance)
+
+
+def test_MCP3001():
+    mock = MockMCP3001(11, 10, 9, 8)
+    with MCP3001() as pot:
```

Comment:
I'm guessing that `pot` implies Potentiometer. IMHO using a single potentiometer to test a differential ADC wouldn't work, so maybe the differential ADC code should use a variable named something other than `pot`, so that the code is easier to read?